src/main/java/build/buildfarm/server/BuildFarmServer.java

```diff
@@ -107,40 +114,67 @@ public BuildFarmServer(
-            .addService(new FetchService(instance))
-            .addService(ProtoReflectionService.newInstance())
-            .addService(new PublishBuildEventService(config.getBuildEventConfig()))
-            .intercept(TransmitStatusRuntimeExceptionInterceptor.instance())
-            .intercept(headersInterceptor)
-            .build();
 
-    logger.log(Level.INFO, String.format("%s initialized", session));
+    serverBuilder.addService(healthStatusManager.getHealthService());
+    serverBuilder.addService(new ActionCacheService(instance, config.getAcPolicy()));
+    serverBuilder.addService(new CapabilitiesService(instance));
+    serverBuilder.addService(
+        new ContentAddressableStorageService(
+            instance, /* deadlineAfter=*/ config.getCasWriteTimeout().getSeconds(), TimeUnit.SECONDS
+            /* requestLogLevel=*/ ));
+    serverBuilder.addService(
+        new ByteStreamService(
+            instance,
+            /* writeDeadlineAfter=*/ config.getBytestreamTimeout().getSeconds(),
+            TimeUnit.SECONDS));
+    serverBuilder.addService(
+        new ExecutionService(
+            instance,
+            config.getExecuteKeepaliveAfterSeconds(),
+            TimeUnit.SECONDS,
+            keepaliveScheduler,
+            getMetricsPublisher(config.getMetricsConfig())));
+    serverBuilder.addService(new OperationQueueService(instance));
+    serverBuilder.addService(new OperationsService(instance));
+    serverBuilder.addService(new AdminService(config.getAdminConfig(), instance));
+    serverBuilder.addService(new FetchService(instance));
+    serverBuilder.addService(ProtoReflectionService.newInstance());
+    serverBuilder.addService(new PublishBuildEventService(config.getBuildEventConfig()));
+
+    serverBuilder.intercept(TransmitStatusRuntimeExceptionInterceptor.instance());
+    serverBuilder.intercept(headersInterceptor);
+    handleGrpcMetricIntercepts(serverBuilder, config);
+
+    return serverBuilder.build();
+  }
+
+  private static void handleGrpcMetricIntercepts(
+      ServerBuilder<?> serverBuilder, BuildFarmServerConfig config) {
+    // Decide how to capture GRPC Prometheus metrics.
+    // By default, we don't capture any.
+    if (config.getGrpcMetrics().getEnabled()) {
+      // Assume cheap metrics.
+      // Cheap metrics include send/receive totals tagged with return codes.  No latencies.
+      Configuration grpcConfig = Configuration.cheapMetricsOnly();
+
+      // Enable latency buckets.
+      if (config.getGrpcMetrics().getProvideLatencyHistograms()) {
+        grpcConfig = grpcConfig.allMetrics();
+      }
+
+      // Enable latency buckets with status codes.
+      if (config.getGrpcMetrics().getTagLatenciesByCode()) {
+        // TODO(luxe):
+        // The feature is not yet included in latest release.
+        // Once the feature is available, we can enable this code:
+        // grpcConfig = grpcConfig.withCodeLabelInLatencyHistogram();
```

Comment:
Safe to leave this in if the default is disabled and this feature is also disabled. I have an aversion to commented-out code.

Reply:
+1. Just put this in a separate branch and reuse it later.

Reply:
removed